### PR TITLE
Add an impersonation test case to the audit E2E test

### LINKED ISSUE #70198: Add an impersonation test case to the audit E2E test
We don't currently have any test coverage of impersonation for audit logging. I think the audit e2e test (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/auth/audit.go) would be a good place to do so.

/area audit
/area test
/sig auth
/help
/good-first-issue

### PR DESCRIPTION
**Pull request by dczombera**
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
Add impersonation test to the audit E2E test 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70198 

**Special notes for your reviewer**:
I have added just one test for the impersonation where I list the pods and check the for the `ImpersonatedUser` field. I wasn't sure if this is enough or if I should add more. 
Also, when parsing the `ImpersonatedUser` struct in `parseAuditLine` I followed the way how it was done with the `User` struct (i.e. only extracting the user name, not the group)

**Does this PR introduce a user-facing change?**: 

```release-note
NONE
```
/sig auth

cc  @tallclair

### CODE REVIEW

**Comment by dczombera** [comment]
I don't quite understand why the e2e tests are failing or how the errors are related to what I did. Does anyone have an idea?

**Comment by tallclair** [comment]
@dczombera that was a tricky one, but I think I figured it out. The CRD error is a red herring, probably because the CRD is failing to be deleted (but we're not checking for it). What is actually happening is that you added a check of the ImpersonatedUser, which is causing the other test cases that set an impersonated user to fail (since they aren't looking for the right event anymore). To fix it, you'll need to set the impersonated user to `system:anonymous` for the anonymous test case:
https://github.com/kubernetes/kubernetes/pull/71209/files#diff-f2cf7c589a05a7b05944e734bbc0ec7cR725

**Comment by liggitt** [comment]
> To fix it, you'll need to set the impersonated user to `system:anonymous` for the anonymous test case:

is there a way to separate the test cases and filter the events we check for to make the test cases more independent?

**Comment by tallclair** [comment]
https://github.com/kubernetes/kubernetes/pull/71489 to address separation

**Comment by dczombera** [comment]
@tallclair Thanks so much for pointing that out. :thumbsup: Sorry that you had to invest your time to find that nasty error with the unset `ImpersonatedUser` field. :disappointed_relieved:
I guess it makes sense to wait until your changes in #71489 are merged (it's my understanding that this will take some time since it's not part of v1.13) before I rebase this branch and adapt the test I added to align with the separation.

**Comment by tallclair** [comment]
Don't worry about #71489, I can rebase mine if yours gets in first.

File: test/utils/audit.go [hunk 1]
```diff
@@ -101,6 +102,9 @@ func parseAuditLine(line string, version schema.GroupVersion) (AuditEvent, error
 	if e.RequestObject != nil {
 		event.RequestObject = true
 	}
+	if e.ImpersonatedUser != nil {
+		event.ImpersonatedUser = e.ImpersonatedUser.Username
+	}
```
Comment (tallclair) [inline test/utils/audit.go]: Do you think we should verify impersonated groups too?

Comment (dczombera) [reply test/utils/audit.go]: Yeah, I think thats a good idea! I extended the tests to also check the groups.
However, as you can see some (seemingly) unrelated tests are failing now.  I ran the tests twice and both times the tests that are failing are different, while `Advanced Audit should audit API calls ` passes both times.

File: test/utils/audit.go [hunk 2]
```diff
@@ -101,6 +104,10 @@ func parseAuditLine(line string, version schema.GroupVersion) (AuditEvent, error
 	if e.RequestObject != nil {
 		event.RequestObject = true
 	}
+	if e.ImpersonatedUser != nil {
+		event.ImpersonatedUser = e.ImpersonatedUser.Username
+		event.ImpersonatedGroups = strings.Join(e.ImpersonatedUser.Groups, ",")
```
Comment (tallclair) [inline test/utils/audit.go]: Hmm, this is going to be flaky if there is more than 1 group. I think you should just sort the groups before the join.

**Comment by tallclair** [comment]
The test failures just look like flakes. Probably unrelated.

File: test/utils/audit.go (hunk 2 above)
Comment (dczombera) [reply test/utils/audit.go]: Good catch. 👍

**Comment by dczombera** [comment]
I just wanted to ask whether there is something else that needs to be fixed before this pr gets merged or if it the current status is fine?

**Comment by liggitt** [comment]
lgtm, please squash and then we can tag

**Comment by dczombera** [comment]
> lgtm, please squash and then we can tag

Done! However, some seemingly unrelated tests fail now (as happened before as well).

**Pull request merged**